### ISSUE TITLE
Make writing "within a test" and "t.XXX" matchers easier:

### ISSUE THREAD

**Issue by jamestalmage** [avajs/eslint-plugin-ava#72]
We have this code construct everywhere:

```js
CallExpression: function (node) {
  // only check if we are within a test file, and if this CallExpression is within a test.
  if (!ava.isTestFile || !ava.currentTestNode) {
    return;
  }
}
```

I propose we add some sugar to `ava.merge()` that allows something like this:

```js
'CallExpression:inTest': function (node) {

},
'VariableDeclaration:inTestFile': function (node) {

},
'CallExpression:inTest:inTestFile': function (node) {

},

```

While we are at it, we should define one (specifically for CallExpression), that checks the callee is a `MemberExpression` of the form `t.XXX`:

```js
'AssertionCallExpression:inTest': function (node) {

}
```

**Comment by jamestalmage** [comment]
Counterpoint: Does our API change often enough that these sorts of helpers would be worth the effort it took to create?

**Comment by sindresorhus** [comment]
// @twada

**Comment by jfmengels** [comment]
Agree with the idea, but I'm not sure it's a good idea. I noticed that we don't handle cases like this
```js
function implementation(t) {
  t.doWhatever();
}

test(implementation);
```
though we might want to, especially with the talk about "macros" in https://github.com/sindresorhus/ava/issues/695. In this case, the "in test" part will pretty much always be wrong.

Otherwise, I think we don't need to add the `inTest` suffix, we could make `ava.isTestFile && ava.currentTestNode` a necessary condition for the visitor to be called (they all check that as far as I can tell).

**Comment by twada** [comment]
Hmm, I think that's out of scope of static analysis tool, it's better and easier to detect and report at runtime.

**Comment by sindresorhus** [comment]
@twada What specifically is out of scope?

**Comment by twada** [comment]
> What specifically is out of scope?

Chasing some kind of macros, renamed variables, destructuring etc..
They tend to make static analyze tools fragile.

**Comment by jfmengels** [comment]
What do you guys think of specifying a simple filter function instead?

From
```js
CallExpression: function (node) {
  // only check if we are within a test file, and if this CallExpression is within a test.
  if (!ava.isTestFile || !ava.currentTestNode) {
    return;
  }
}
```
to
```js
// exported by create-ava-rule for instance
function ifInAVATest(fn) {
  return function(node) {
    if (ava.isTestFile && ava.currentTestNode) {
      return fn(node);
    }
  };
}

// test file
CallExpression: ifInAVATest(function (node) {
  // do stuff
})
```

No new sugar, just explicit functions, and they're chainable too if needed.

**Comment by jfmengels** [comment]
Maybe just have some predicate functions and a visitor filter function

```js
// create-ava-rule.js
function isAssertion(node) {
  return node.type === 'CallExpression' && ...
}

function callIf(...fns) {
  return function(visitor) {
    return function (node) {
      // some kind of
      if (fns.every((fn) => fn(node)) {
        visitor(node);
      }
    };
  };
}

// test file
CallExpression: callIf(ava.isAssertion, ava.fooBar)(function (node) {
  // do stuff
  // I can still use ava.isAssertion programatically
})
```

Kind of messy syntax, but I hope you get the gist

**Comment by sindresorhus** [comment]
@jfmengels I like it. @jamestalmage ?

**Comment by twada** [comment]
@jfmengels Wow it looks nice :)

**Comment by jamestalmage** [comment]
:+1:
